dipy/reconst/qti.py

```diff
@@ -0,0 +1,652 @@
+    for i, bten in enumerate(gtab.btens):
+        bten = from_3x3_to_6x1(bten)
+        bten_sq = from_6x6_to_21x1(np.matmul(bten, np.swapaxes(bten, -2, -1)))
+        S[i] = S0 * np.exp(
+            - np.matmul(np.swapaxes(bten, -2, -1), D)
+            + .5 * np.matmul(np.swapaxes(bten_sq, -2, -1), C))
+    return S
+
+
+def design_matrix(btens):
+    """Calculate the QTI design matrix from the b-tensors.
+
+    Parameters
+    ----------
+    btens : numpy.ndarray
+        An array of b-tensors of shape (number of acquisitions, 3, 3).
+
+    Returns
+    -------
+    X : numpy.ndarray
+        QTI design matrix.
+    """
+    X = np.zeros((btens.shape[0], 28))
+    for i, bten in enumerate(btens):
+        b = from_3x3_to_6x1(bten)
+        b_sq = from_6x6_to_21x1(np.matmul(b, b.T))
+        X[i] = np.concatenate(
+            ([1], (-b.T)[0, :], (0.5 * b_sq.T)[0, :]))
+    return X
+
+
+def _ols_fit(data, mask, X):
+    """Estimate the QTI model parameters using ordinary least squares.
+
+    Parameters
+    ----------
+    data : numpy.ndarray
+        Array of shape (..., number of acquisitions).
+    mask : numpy.ndarray
+        Array with the same shape as the data array of a single acquisition.
+    X : numpy.ndarray
+        Design matrix of shape (number of acquisitions, 28).
+
+    Returns
+    -------
+    params : numpy.ndarray
+        Array of shape (..., 28) containing the estimated model parameters.
+        Element 0 is the estimated signal without diffusion-weighting, elements
+        1-6 are the estimated diffusion tensor parameters, and elements 7-28 are
+        the estimated covariance tensor parameters.
+    """
+    params = np.zeros(mask.shape + (28,)) * np.nan
+    X_inv = np.linalg.pinv(np.matmul(X.T, X))  # Independent of data
+    index = ndindex(mask.shape)
+    for v in index:  # This loop is slow
+        if not mask[v]:
+            continue
+        S = np.log(data[v])[:, np.newaxis]
+        params[v] = np.matmul(X_inv, np.matmul(X.T, S))[:, 0]
+    return params
+
+
+def _wls_fit(data, mask, X):
+    """Estimate the QTI model parameters using weighted least squares where the
+    signal magnitudes are used as the weights.
+
+    Parameters
+    ----------
+    data : numpy.ndarray
+        Array of shape (..., number of acquisitions).
+    mask : numpy.ndarray
+        Array with the same shape as the data array of a single acquisition.
+    X : numpy.ndarray
+        Design matrix of shape (number of acquisitions, 28).
+
+    Returns
+    -------
+    params : numpy.ndarray
+        Array of shape (..., 28) containing the estimated model parameters.
+        Element 0 is the estimated signal without diffusion-weighting, elements
+        1-6 are the estimated diffusion tensor parameters, and elements 7-27 are
+        the estimated covariance tensor parameters.
+    """
+    params = np.zeros(mask.shape + (28,)) * np.nan
+    index = ndindex(mask.shape)
+    for v in index:  # This loop is slow loop
```

Comment:
Check out https://github.com/dipy/dipy/blob/master/dipy/reconst/dti.py#L1224 for our approach there. 

See also: https://github.com/dipy/dipy/blob/master/dipy/reconst/dti.py#L1442 for the solve itself.

And check out https://github.com/dipy/dipy/blob/c846bf5a23b7e95343a9cf231df2653473602456/dipy/utils/arrfuncs.py#L29 for vectorized pinv

Reply:
@kerkelae : might also be worthwhile to compare the above `pinv` with the SciPy `pinv2` both seem to be doing something similar. Also, the SciPy SVD is in general faster than NumPy. Also, it allows you to set `check_finite=False`. 

https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.pinv2.html

Maybe we should change the SVD in `arrfuncs.py` too?